argus/action_manager/windows.py

```diff
@@ -115,34 +117,95 @@ def get_installation_script(self):
+
+    def cbinit_cleanup(self):
+        """Clean-up Cloudbase-Init if it's installed."""
+        LOG.info("Cleaning up Cloudbase-Init.")
+
+        try:
+            cbinit_dir = introspection.get_cbinit_dir(
+                lambda cmd, command_type: self._client.run_command_with_retry(
+                    cmd, count=util.RETRY_COUNT,
+                    delay=util.RETRY_DELAY, command_type=command_type)[0]
+            )
+        except exceptions.ArgusError as exc:
+            LOG.warning("Could not find Cloudbase-Init directory: %s", exc)
+        else:
+            self.rmdir(ntpath.dirname(cbinit_dir))
+
     def install_cbinit(self):
-        """Run the installation script for CloudbaseInit."""
-        LOG.debug("Installing Cloudbase-Init ...")
+        """Install Cloudbase-Init on the underlying instance."""
+        LOG.info("Trying to install Cloudbase-Init ...")
+
+        for _ in range(util.RETRY_COUNT):
+            for install_method in (self._run_installation_script,
+                                   self._deploy_using_scheduled_task):
+                try:
+                    install_method()
+                except exceptions.ArgusError as exc:
+                    LOG.debug("Could not install Cloudbase-Init: %s", exc)
+                    self.cbinit_cleanup()
+                else:
+                    if self.check_cbinit_installation():
+                        return
+                    else:
+                        break
+
+    def _run_installation_script(self):
+        """Run the installation script for Cloudbase-Init."""
+        LOG.info("Running the installation script for Cloudbase-Init ...")
 
         installer = "CloudbaseInitSetup_{build}_{arch}.msi".format(
             build=self._conf.argus.build,
             arch=self._conf.argus.arch
         )
-        # TODO(cpopa): the service type is specific to each scenario,
-        # find a way to pass it
-        LOG.info("Run the downloaded installation script "
-                 "using the installer %r.", installer)
-
         parameters = '-installer {}'.format(installer)
-        try:
-            self.execute_powershell_resource_script(
-                resource_location='windows/installCBinit.ps1',
-                parameters=parameters)
-        except exceptions.ArgusError:
-            # This can happen for multiple reasons,
-            # but one of them is the fact that the installer
-            # can't be installed through WinRM on some OSes
-            # for whatever reason. In this case, we're falling back
-            # to use a scheduled task.
-            LOG.debug("Cannot install, deploying using a scheduled task.")
-            self._deploy_using_scheduled_task(installer)
-
-    def _deploy_using_scheduled_task(self, installer):
+        self.execute_powershell_resource_script(
+            resource_location='windows/installCBinit.ps1',
+            parameters=parameters)
+
+    def _deploy_using_scheduled_task(self):
```

Comment:
I think it was alright to simply give the already created `installer` string from the `_run_installation_script` since you already had it formatted there.